Feature/38 add admin

Leehunil opened the pull request:
resolved : #38 
## 📌 과제 설명
어드민 관련 API 작성

## 👩‍💻 요구 사항과 구현 내용
- 게시글의 승인 상태를 조회 조건에 추가해서 페이징 처리하여 게시글 목록을 조회하였습니다.
- 수정 게시글과 수정, 삭제 게시글의 상세 정보 형태가 달라서 따로 API를 구현하였습니다.
- 게시글 반영 API를 작성하였습니다.
- 게시글 반영은 추가, 수정, 삭제 이렇게 게시글 유형마다 다르게 로직이 진행되게 하였습니다.
- securityConfig에서 hasRole은 prefix에 ROLE을 자동으로 붙여 비교하기 어려워 hasAuthority로 변경하였습니다.

## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->
- 게시글 유형마다 다른 로직을 실행하는 부분을 if문에서 switch문으로 변경
- 디버깅을 용이하게 하기 위해 default는 예외를 던짐

## ✅ PR 포인트 & 궁금한 점
- 게시글 반영하는 부분의 로직에 집중해서 보시면 좋을 것 같아요

Leehunil committed:
feat : Board 상태 변경 로직

- 승인 상태 변경 로직 작성
- TrashcanId 업데이트 로직 작성
- Board를 통해 Trashcan Entity 생성하기

Leehunil committed:
feat : Trashcan Entity 상태 변경 로직

- 쓰레기통 상태 update 로직 작성
- 쓰레기통 img url update 로직 작성

Leehunil committed:
feat : Trashcan Service 로직 변경

- admin service에서 사용할 getTrashcanById를 public으로 변경
- admin service에서 사용할 saveTrashcan 메서드 생성

Leehunil committed:
feat : BoardRepository 쿼리 추가

- approvalStatus, boardCategory, trashCategory로 Boards 조회 쿼리 추가
- boardId에 해당하는 imgUrl이 있는지 확인하는 쿼리 추가

Leehunil committed:
feat : BoardService 로직 추가

- approvalStatus, boardCategory, trashCategory로 Boards 조회해서 페이징
- boardId에 해당하는 imgUrl이 있는지 확인하는 로직 추가

Leehunil committed:
feat : admin 관련 로직 작성

- 관리자 게시글 리스트 API
- 관리자 수정 신고 게시글 조회 API
- 관리자 등록 및 삭제 신고 게시글 조회 API
- 관리자 신고 게시글 반영 API

Leehunil committed:
feat : SecurityConfig 설정 변경

- hasRole은 prefix에 ROLE을 자동을 붙이기 때문에 hasAuthority로 변경

l2yujw commented on src/main/java/com/sscanner/team/admin/responsedto/AdminBoardInfoResponseDTO.java:
```diff
@@ -0,0 +1,22 @@
+        String detailedAddress,
+        ApprovalStatus approvalStatus
+) {
+    public static AdminBoardInfoResponseDTO of(Board board, String boardFirstImgUrl) {
```
of랑 from을 찾아보니까 
다른 객체로 부터 DTO를 생성하면 from을 쓰고
일반적은 변수값들을 받아서 새로운 DTO를 생성하면 of를 쓴다고 하더라고요
from 쓰시는거 어떠실까요?!

Leehunil replied:
정원님 덕분에 of와 from의 차이를 정확히 이해했습니다!! 다만 저희 코드 컨벤션에는 매개변수 개수로 of와 from을 구분하더라고요..!! 이 부분도 나중에 같이 고민해보면 좋을 것 같습니다~ 감사합니다

l2yujw commented on src/main/java/com/sscanner/team/admin/service/AdminServiceImpl.java:
```diff
@@ -0,0 +1,139 @@
+
+    private void processBoardApproval(ApprovalStatus approvalStatus, Board board, String chosenImgUrl) {
+        if(approvalStatus.equals(ApprovalStatus.APPROVED)) {
+            BoardCategory boardCategory = board.getBoardCategory();
```
이 부분 메서드 분리하고 switch(category) 형식을 쓰는건 어떤가요?
뭐가 더 좋은지는 몰라서 의견 드려봅니다!

Leehunil replied:
조건이 복잡하지 않은 단순 문자열 비교라서 switch문을 활용하면 더 좋을 것 같아요!! 감사합니다~

Leehunil committed:
fix : if문에서 switch case문으로 변경

- 단순 문자열 비교는 switch case문이 더 가독성이 좋아보여 바꾸었다.
- switch case문의 default에 예외를 던져 잘못된 게시글 유형이 들어왔을 경우 디버깅이 용이하게 하였습니다.
